Zend/zend_strictmap.c

```diff
@@ -0,0 +1,243 @@
+//
+// 	const uint32_t nIndex = h | array->hash_mask;
+// 	uint32_t i = SM_HASH(array, nIndex);
+//
+// 	const uint32_t idx = entry - entries;
+// 	if (i == idx) {
+// 		SM_HASH(array, nIndex) = SM_IT_NEXT(entry);
+// 	} else {
+// 		zend_strictmap_entry *prev = &array->data[i];
+// 		while (SM_IT_NEXT(prev) != idx) {
+// 			i = SM_IT_NEXT(prev);
+// 			prev = &array->data[i];
+// 		}
+// 		SM_IT_NEXT(prev) = SM_IT_NEXT(entry);
+// 	}
+//
+// 	array->count--;
+//
+// 	SM_IT_NEXT(entry) = array->free_index_chain;
+// 	array->free_index_chain = idx;
+//
+// 	if (array->first_used_index == idx) {
+// 		if (array->count == 0) {
+// 			ZEND_ASSERT(array->used_indices == 0);
+// 			array->first_used_index = 0;
+// 		} else {
+// 			ZEND_ASSERT(array->first_used_index < array->used_indices);
+// 			do {
+// 				array->first_used_index++;
+// 			} while (array->first_used_index < array->used_indices && (UNEXPECTED(Z_ISUNDEF(array->data[array->first_used_index].key))));
+// 			ZEND_ASSERT(array->first_used_index < array->used_indices);
+// 		}
+// 	}
+// 	zval old_key;
+// 	zval old_value;
+// 	ZVAL_COPY_VALUE(&old_key, &entry->key);
+// 	ZVAL_COPY_VALUE(&old_value, &entry->value);
+// 	ZEND_ASSERT(array->count <= array->used_indices);
+// 	ZEND_ASSERT(array->first_used_index < array->used_indices || (array->first_used_index == array->used_indices && array->count == 0));
+// 	ZVAL_UNDEF(&entry->key);
+// 	ZVAL_UNDEF(&entry->value);
+//
+// 	return true;
+// }
+
+static size_t zend_strictmap_offset_bytes_for_capacity(uint32_t capacity) {
+	return capacity * sizeof(uint32_t);
+}
+
+static zend_strictmap_entry *zend_strictmap_alloc_entries(uint32_t capacity) {
+	uint8_t *ptr = safe_emalloc(capacity, SM_MEMORY_PER_ENTRY, 0);
+	const size_t buckets_byte_count = zend_strictmap_offset_bytes_for_capacity(capacity);
+	memset(ptr, SM_INVALID_INDEX, buckets_byte_count);
+	return (void *)(ptr + buckets_byte_count);
+}
+
+static void zend_strictmap_free_entries(zend_strictmap_entry *old_entries, uint32_t old_capacity) {
+	void * old_ptr = ((uint8_t *) old_entries) - zend_strictmap_offset_bytes_for_capacity(old_capacity);
+	efree(old_ptr);
+}
+
+static void zend_strictmap_grow(zend_strictmap *array)
+{
+	ZEND_ASSERT(array->used_indices >= array->count);
+	if (UNEXPECTED(array->capacity >= HT_MAX_SIZE)) {
+		zend_error_noreturn(E_ERROR, "exceeded max valid strictmap capacity");
+	}
+
+	const uint32_t new_capacity = array->capacity * 2;
+	zend_strictmap_entry *const orig_entries = array->data;
+	zend_strictmap_entry *const new_entries = zend_strictmap_alloc_entries(new_capacity);
+	zend_strictmap_entry *old_entry;
+	zend_strictmap_entry *it = new_entries;
+	ZEND_ASSERT(array->count <= array->used_indices);
+	uint32_t i = 0;
+	const uint32_t new_mask = SM_SIZE_TO_MASK(new_capacity);
+	ZEND_STRICTMAP_FOREACH_BUCKET(array, old_entry) {
+		/* TODO: This can be optimized based on zend_hash_rehash's special cases, e.g. HT_IS_WITHOUT_HOLES */
```

Comment:
@TysonAndre Can you clarify on how this could be optimized? I don't understand what else we could do here.

Reply:
I'd just remove the todo and not bother doing that. It's a longer term note for StrictHashSet/StrictHashMap for a micro-optimization.

It's a note on how zend_hash_extend would memcpy the buckets directly when growing a hash map, and how zend_hash_rehash didn't need to do anything special to the memcpyed data when HT_IS_WITHOUT_HOLES is tracked (though there is no memcpy here yet)

- i.e. the ZVAL_COPY_VALUE 

I didn't even start tracking HT_IS_WITHOUT_HOLES (though in this use case, it's guaranteed not to have holes)